:sparkles: Add go/v4-alpha base and migration guide from go/v3 to go/v4-alpha (#3032)
kubernetes-sigs/kubebuilder: docs/book/src/migration/v3vsv4.md

# go/v3 vs go/v4-alpha 

> This document covers all breaking changes when migrating from v2 to v3.

Reviewer: Is this `from v3 to v4`?

> The details of all changes (breaking or otherwise) can be found in

Reviewer: Nits:

```
The details of all changes (breaking or otherwise) can be found in:
- [controller-runtime][controller-runtime]
- [controller-tools][controller-tools]
- [kustomize][kustomize-release]
- [kb-releases][kb-releases] release notes.
```

[controller-runtime][controller-runtime],
[controller-tools][controller-tools], [kustomize][kustomize-release]
and [kb-releases][kb-releases] release notes.

## Common changes

> - `go/v4-alpha` projects use Kustomize v4x (instead of v3x)

Reviewer: Nits:

use Kustomize `v4.x.x` (instead of `v3.x.x`)

Reviewer: That makes sense. 👍🏼

# go/v3 vs go/v4-alpha 

This document covers all breaking changes when migrating from projects built using the plugin go/v3 (default for any scaffold done since `28 Apr 2021`) to the next alpha version of the Golang plugin `go/v4-alpha`.

The details of all changes (breaking or otherwise) can be found in:
- [controller-runtime][controller-runtime]
- [controller-tools][controller-tools]
- [kustomize][kustomize-release]
- [kb-releases][kb-releases] release notes.

## Common changes

- `go/v4-alpha` projects use Kustomize v4x (instead of v3x)

> - note that some manifests under `config/` directory changes in order to no longer use the deprecated Kustomize features

Reviewer 2: ```suggestion
- note that some manifests under `config/` directory have been changed in order to no longer use the deprecated Kustomize features
```

such as env vars. 
- adds support for Apple Silicon M1 (darwin/arm64)
- remove support to CRD/WebHooks Kubernetes API v1beta1 version which are no longer supported since k8s 1.22
- no longer scaffold webhook test files with `"k8s.io/api/admission/v1beta1"` the k8s API which is no longer served since k8s `1.25`. By default
webhooks test files are scaffolding using `"k8s.io/api/admission/v1"` which is support from k8s `1.20`
- no longer provide backwards compatible support with k8s versions < `1.16`

<aside class="note">

> <H1> TL;DR of the New `go/v4-alpha` Plugin </H1>

Reviewer: See that below contains another `tldr` header with contents. Do we need this `h1` block?

Author: All notes have that. It creates a title @Kavinjsir

Further details can be found in the [go/v4-alpha plugin section][go/v4-doc]

</aside>

## TL;DR of the New `go/v4-alpha` Plugin

***More details on this can be found at [here][kb-releases], but for the highlights, check below***

<aside class="note">
<h1>Default plugin</h1>
Projects scaffolded with Kubebuilder v3 release still using the `go.kubebuilder.io/v3` plugin by default.

</aside>

<aside class="note warning">
<h1>Project customizations</h1>

After using the CLI to create your project, you are free to customise how you see fit. Bear in mind, that it is not recommended to deviate from the proposed layout unless you know what you are doing.

For example, you should refrain from moving the scaffolded files, doing so will make it difficult in upgrading your project in the future. You may also lose the ability to use some of the CLI features and helpers. For further information on the project layout, see the doc [What's in a basic project?][basic-project-doc]

</aside>

## Migrating to Kubebuilder go/v4-alpha

> So you want to upgrade your scaffolding to use the latest and greatest features then, follow up the following guide 

Reviewer 2: ```suggestion
If you want to upgrade your scaffolding to use the latest and greatest features then, follow the guide 
```

which will cover the steps in the most straightforward way to allow you to upgrade your project to get all
latest changes and improvements.

<aside class="note warning">
<h1> Apple Silicon (M1) </h1>

The current scaffold done by the CLI (`go/v3`) uses [kubernetes-sigs/kustomize][kustomize] v3 which does not provide
a valid binary for Apple Silicon (`darwin/arm64`). Therefore, you can use the `go/v4-alpha` plugin
instead which provides support for this platform:

```bash
kubebuilder init --domain my.domain --repo my.domain/guestbook --plugins=go/v4-alpha
```

**Note**: The `go/v4-alpha` plugin is an unstable version and can have breaking changes in future releases.
</aside>

- [Migration Guide go/v3 to go/v4][migration-guide-gov3-to-gov4] **(Recommended)**

### By updating the files manually

> So you want to use the latest version of Kubebuilder CLI without changing your scaffolding then, check the following guide which will describe the manually steps required for you to upgrade only your PROJECT version and starts to use the plugins versions.

Reviewer 2: ```suggestion
If you want to use the latest version of Kubebuilder CLI without changing your scaffolding then, check the following guide which will describe the steps to be performed manually to upgrade only your PROJECT version and start using the plugins versions.
```
